File: coconut/src/rsync.rs
use std::fs::create_dir_all;
use std::io::{Error as IoError, ErrorKind as IoErrorKind};
use std::path::{Path, PathBuf};
use std::process::Command;

use clap::Parser;
use palm::{print_command_output, timestamp_file, Result};

#[derive(Parser, PartialEq, Eq, Debug, Clone)]
pub struct Config {
    #[arg(short, long, default_value = super::Args::DEFAULT_KEEP)]
    pub keep: usize,
    #[arg(short, long, default_value = super::Args::DEFAULT_TARGET)]
    pub target: PathBuf,
    #[arg(short = 'H', long, help = "Keep it empty if backup from local")]
    pub host: Option<String>,
    #[arg(short, long, default_value = "22")]
    pub port: u16,
    #[arg(short, long)]
    pub source: PathBuf,
    #[arg(short, long, default_value = "root")]
    pub user: String,
    #[arg(short = 'P', long)]
    pub password: Option<String>,
    #[arg(short = 'K', long)]
    pub key: Option<PathBuf>,
}

impl Config {
    pub fn launch<P: AsRef<Path>>(&self, target: P) -> Result<String> {
        let target = target.as_ref();

        let name = timestamp_file(&self.user, None);
        let tmp = target.join(&name);
        create_dir_all(target)?;

        let out = match self.host {
            Some(ref host) => {
                info!(
                    "backup ssh://{}@{}:{}:{} to {}",
                    self.user,
                    host,
                    self.port,
                    self.source.display(),
                    target.display()
                );
                match self.password {
                    Some(ref password) => Command::new("sshpass")
                        .arg("-p")
                        .arg(password)
                        .arg("rsync")
                        .arg("-a")
                        .arg("-z")
                        .arg("-q")
                        .arg("-e")
                        .arg(format!(
                            "ssh -o StrictHostKeyChecking=yes -p {} -l {}",
                            self.port, self.user,
                        ))
                        .arg(&format!("{}:{}", host, self.source.display()))
                        .arg(&tmp)
                        .output(),
                    None => Command::new("rsync")
                        .arg("-a")
                        .arg("-z")
                        .arg("-q")
                        .arg("-e")
                        .arg(format!(
                            "ssh -o StrictHostKeyChecking=yes -p {} -l {} -i {}",
                            self.port,
                            self.user,
                            self.ssh_key_file()?.display(),
                        ))
                        .arg(&format!("{}:{}", host, self.source.display()))
                        .arg(&tmp)
                        .output(),
                }
            }
            None => {
                info!("backup {} to {}", self.source.display(), target.display());
                Command::new("cp")
                    .arg("-a")
                    .arg(&self.source)
                    .arg(&tmp)
                    .output()
            }
        }?;

        print_command_output(&out)?;
        Ok(name)
    }

    fn ssh_key_file(&self) -> Result<PathBuf> {
        if let Some(ref it) = self.key {
            return Ok(it.clone());
        }
        if let Some(ref home) = dirs::home_dir() {
            let home = home.join(".ssh");
            for it in ["id_ed25519", "id_rsa"] {
                let it = home.join(it);
                if it.is_file() {
                    return Ok(it);
                }
            }
        }
        Err(Box::new(IoError::from(IoErrorKind::UnexpectedEof)))
    }
}

File: coconut/coconut/src/rsync.cpp
#include "coconut/rsync.hpp"
#include "coconut/nut.hpp"

coconut::Rsync::Rsync(const toml::table& root) {
  this->host = root["host"].value<std::string>();
  this->port = root["port"].value_or(22);
  this->user = root["user"].value_or("root");
  this->password = root["password"].value<std::string>();
  this->source = root["source"].value<std::string>().value();

  {
    std::optional<std::string> key = root["key"].value<std::string>();
    this->set_key(key);
  }
}
void coconut::Rsync::set_key(std::optional<std::string> key) {
  if (key) {
    this->key = key.value();
  } else {
    const auto home = coconut::home(".ssh");
    for (const auto& it : {"id_ed25519", "id_rsa"}) {
      const auto file = home / it;
      if (std::filesystem::is_regular_file(file)) {
        this->key = file;
        break;
      }
    }
  }
}

void coconut::Rsync::dump(const std::filesystem::path& target,
                          const size_t keep) const {
  const std::string cur = std::filesystem::path(this->source).stem().string() +
                          "-" + coconut::timestamp();
  const auto root = target / cur;
  {
    const std::string it = target.string();
    spdlog::debug("change workdir to {}", it);
    const auto ok = chdir(it.c_str());
    if (ok != EXIT_SUCCESS) {
      throw std::runtime_error("switch to folder " + it);
    }
  }

  if (this->host) {
    const auto host = this->host.value();
    std::stringstream ss;

    if (this->password) {
      ss << "sshpass -p " << this->password.value()
         << " rsync -a -z -q -e 'ssh -o StrictHostKeyChecking=yes -p "
         << this->port << " -l " << this->user << "' ";
    } else {
      ss << "rsync -a -z -q -e 'ssh -o StrictHostKeyChecking=yes -p "
         << this->port << " -l " << this->user << " -i "
         << this->key.value_or("id_rsa").string() << "' ";
    }
    ss << host << ":" << this->source << " " << root.string();
    const auto cmd = ss.str();
    spdlog::debug("{}", cmd);
    const auto ok = std::system(cmd.c_str());
    if (ok != EXIT_SUCCESS) {
      throw std::runtime_error("backup " + host + ":" + this->source);
    }
  } else {
    spdlog::info("copy folder {} to {}", this->source, root.string());
    std::filesystem::copy(this->source, root);
  }

  coconut::tar(cur);
  coconut::keep(target, keep);
}
